src/ducc0/infra/threading.h

```diff
@@ -53,21 +53,106 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #ifndef DUCC0_THREADING_H
 #define DUCC0_THREADING_H
 
+// Low level threading support can be influenced by the following macros:
+// - DUCC0_NO_LOWLEVEL_THREADING: if defined, multithreading is disabled
+//   and all parallel regions will be executed sequentially
+//   on the invoking thread.
+// - DUCC0_CUSTOM_LOWLEVEL_THREADING: if defined, external definitions of
+//   Mutex, UniqueLock, LockGuard, CondVar, thread_pool, set_active_pool(),
+//   and get active_pool() must be supplied,
+//   and the code will use those.
+// Both macros must not be defined at the same time.
+// If neither macro is defined, standard ducc0 multihreading will be active.
+
+#if (defined(DUCC0_NO_LOWLEVEL_THREADING) && defined(DUCC0_CUSTOM_LOWLEVEL_THREADING))
+static_assert(false, "DUCC0_NO_LOWLEVEL_THREADING and DUCC0_CUSTOMLOWLEVEL_THREADING must not be both defined");
+#endif
+
+#if defined(DUCC0_STDCXX_LOWLEVEL_THREADING)
+static_assert(false, "DUCC0_STDCXX_LOWLEVEL_THREADING must not be defined externally");
+#endif
+
+#if ((!defined(DUCC0_NO_LOWLEVEL_THREADING)) && (!defined(DUCC0_CUSTOM_LOWLEVEL_THREADING)))
+#define DUCC0_STDCXX_LOWLEVEL_THREADING
+#endif
+
 #include <cstddef>
 #include <functional>
-#ifndef DUCC0_NO_THREADING
+#include <optional>
+
+// threading-specific headers
+#ifdef DUCC0_STDCXX_LOWLEVEL_THREADING
 #include <mutex>
 #include <condition_variable>
 #endif
-#include <optional>
+
+#ifdef DUCC0_NO_LOWLEVEL_THREADING
+// no headers needed
+#endif
+
+#ifdef DUCC0_CUSTOM_LOWLEVEL_THREADING
+// include headers necessary for custom threading support here.
+#endif
 
 namespace ducc0 {
 
 namespace detail_threading {
 
+class thread_pool;
+
+thread_pool *set_active_pool(thread_pool *new_pool);
+thread_pool *get_active_pool();
+
+// define threading related types dependent on the underlying implementation
+#ifdef DUCC0_STDCXX_LOWLEVEL_THREADING
+using Mutex = std::mutex;
+using UniqueLock = std::unique_lock<std::mutex>;
+using LockGuard = std::lock_guard<std::mutex>;
+using CondVar = std::condition_variable;
+#endif
+
+#ifdef DUCC0_NO_LOWLEVEL_THREADING
+struct Mutex
+  {
+  void lock(){}
+  void unlock(){}
+  };
+struct LockGuard
+  {
+  LockGuard(const Mutex &){}
+  };
+struct UniqueLock
+  {
+  UniqueLock(const Mutex &){}
+  };
+struct CondVar
+  {
+  template<class Predicate>
+    void wait(UniqueLock &, Predicate) {}
+  void notify_one() noexcept {}
+  void notify_all() noexcept {}
+  };
+#endif
+
+#ifdef DUCC0_CUSTOM_LOWLEVEL_THREADING
```

Comment:
Ideally a user wouldn't need to actually modify your source files.  If there was a configuration header included above, the definitions would be contained within that.

Reply:
OK.